prow/gitoauth/gitoauth.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gitoauth
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"golang.org/x/net/xsrftoken"
+	"golang.org/x/oauth2"
+	"k8s.io/test-infra/prow/config"
+)
+
+const (
+	oauthSessionCookie = "oauth-session"
+	stateKey           = "state"
+)
+
+// GitOAuth Agent represents an agent that takes care Github authentication process such as handles
+// login request from users or handles redirection from Github OAuth server.
+type GitOAuthAgent struct {
+	gc     *config.GitOAuthConfig
+	logger *logrus.Entry
+}
+
+// Returns new GitOAUth Agent.
+func NewGitOAuthAgent(config *config.GitOAuthConfig, logger *logrus.Entry) *GitOAuthAgent {
+	return &GitOAuthAgent{
+		gc:     config,
+		logger: logger,
+	}
+}
+
+// HandleLogin handles Github login request from front-end. It starts a new git oauth session and
+// redirect user to Github OAuth end-point for authentication.
+func (ga *GitOAuthAgent) HandleLogin(w http.ResponseWriter, r *http.Request) {
+	state := xsrftoken.Generate(ga.gc.Client.ClientSecret, "", "")
+
+	oauthSession, err := ga.gc.CookieStore.New(r, oauthSessionCookie)
+	oauthSession.Options.MaxAge = 10 * 60
```

Comment:
If oauthSession is nil, this will panic. Check the error first before setting this.

Reply:
done